File: backend/app/models/custom_resume.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime
from app.models.heading import HeadingResponse
from app.models.education import EducationResponse
from app.models.experience import ExperienceResponse
from app.models.project import ProjectResponse
from app.models.skill import SkillResponse
from app.models.volunteer import VolunteerResponse
from app.models.certification import CertificationResponse
from app.models.award import AwardResponse

# Request Models
class CustomResumeCreate(BaseModel):
    name: str
    heading_ids: Optional[List[str]] = []
    education_ids: Optional[List[str]] = []
    experience_ids: Optional[List[str]] = []
    project_ids: Optional[List[str]] = []
    skill_ids: Optional[List[str]] = []
    volunteer_ids: Optional[List[str]] = []
    certification_ids: Optional[List[str]] = []
    award_ids: Optional[List[str]] = []

class CustomResumeUpdate(BaseModel):
    name: Optional[str] = None
    heading_ids: Optional[List[str]] = None
    education_ids: Optional[List[str]] = None
    experience_ids: Optional[List[str]] = None
    project_ids: Optional[List[str]] = None
    skill_ids: Optional[List[str]] = None
    volunteer_ids: Optional[List[str]] = None
    certification_ids: Optional[List[str]] = None
    award_ids: Optional[List[str]] = None

# Response Models
class CustomResumeResponse(BaseModel):
    id: str
    user_id: str
    name: str
    headings: List[HeadingResponse] = []
    educations: List[EducationResponse] = []
    experiences: List[ExperienceResponse] = []
    projects: List[ProjectResponse] = []
    skills: List[SkillResponse] = []
    volunteers: List[VolunteerResponse] = []
    certifications: List[CertificationResponse] = []
    awards: List[AwardResponse] = []
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class UserElementsResponse(BaseModel):
    headings: List[HeadingResponse] = []
    educations: List[EducationResponse] = []
    experiences: List[ExperienceResponse] = []
    projects: List[ProjectResponse] = []
    skills: List[SkillResponse] = []
    volunteers: List[VolunteerResponse] = []
    certifications: List[CertificationResponse] = []
    awards: List[AwardResponse] = []
    
    class Config:
        from_attributes = True

File: frontend/src/api/custom-resume.ts
import api from './axios'
import type { HeadingResponse } from './heading'
import type { EducationResponse } from './education'
import type { ExperienceResponse } from './experience'
import type { ProjectResponse } from './project'
import type { SkillResponse } from './skill'
import type { VolunteerResponse } from './volunteer'
import type { CertificationResponse } from './certification'
import type { AwardResponse } from './award'

export interface CustomResume {
  id: string
  user_id: string
  name: string
  headings: HeadingResponse[]
  educations: EducationResponse[]
  experiences: ExperienceResponse[]
  projects: ProjectResponse[]
  skills: SkillResponse[]
  volunteers: VolunteerResponse[]
  certifications: CertificationResponse[]
  awards: AwardResponse[]
  created_at: string
  updated_at: string
}

export interface CustomResumeCreate {
  name: string
  heading_ids?: string[]
  education_ids?: string[]
  experience_ids?: string[]
  project_ids?: string[]
  skill_ids?: string[]
  volunteer_ids?: string[]
  certification_ids?: string[]
  award_ids?: string[]
}

export interface CustomResumeUpdate {
  name?: string
  heading_ids?: string[]
  education_ids?: string[]
  experience_ids?: string[]
  project_ids?: string[]
  skill_ids?: string[]
  volunteer_ids?: string[]
  certification_ids?: string[]
  award_ids?: string[]
}

export interface UserElementsResponse {
  headings: HeadingResponse[]
  educations: EducationResponse[]
  experiences: ExperienceResponse[]
  projects: ProjectResponse[]
  skills: SkillResponse[]
  volunteers: VolunteerResponse[]
  certifications: CertificationResponse[]
  awards: AwardResponse[]
}

export const customResumeApi = {
  getAllResumes: async (): Promise<CustomResume[]> => {
    const response = await api.get<CustomResume[]>('/api/custom-resume/')
    return response.data
  },

  getResumeById: async (resumeId: string): Promise<CustomResume> => {
    const response = await api.get<CustomResume>(`/api/custom-resume/${resumeId}`)
    return response.data
  },

  createResume: async (data: CustomResumeCreate): Promise<CustomResume> => {
    const response = await api.post<CustomResume>('/api/custom-resume/', data)
    return response.data
  },

  updateResume: async (resumeId: string, data: CustomResumeUpdate): Promise<CustomResume> => {
    const response = await api.put<CustomResume>(`/api/custom-resume/${resumeId}`, data)
    return response.data
  },

  deleteResume: async (resumeId: string): Promise<void> => {
    await api.delete(`/api/custom-resume/${resumeId}`)
  },

  getUserElements: async (): Promise<UserElementsResponse> => {
    const response = await api.get<UserElementsResponse>('/api/custom-resume/user-elements')
    return response.data
  },
}
